sycl/include/CL/sycl/property_list.hpp

```diff
@@ -8,303 +8,96 @@
 
 #pragma once
 
-#include <CL/sycl/context.hpp>
-#include <CL/sycl/detail/common.hpp>
-#include <tuple>
+#include <CL/sycl/detail/property_base.hpp>
+#include <CL/sycl/exception.hpp>
+#include <memory>
 #include <type_traits>
-#include <utility>
+#include <typeindex>
+#include <unordered_map>
 
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
 
-// HOW TO ADD NEW PROPERTY INSTRUCTION:
-// 1. Add forward declaration of property class.
-// 2. Add new record in PropKind enum.
-// 3. Use RegisterProp macro passing new record from enum and new class.
-// 4. Add implementation of the new property class using detail::Prop class with
-//    template parameter = new record in enum as a base class.
-
-namespace property {
-
-namespace image {
-class use_host_ptr;
-class use_mutex;
-class context_bound;
-} // namespace image
-
-namespace buffer {
-class use_host_ptr;
-class use_mutex;
-class context_bound;
-} // namespace buffer
-
-namespace queue {
-class enable_profiling;
-class in_order;
-} // namespace queue
-
-class noinit;
-
-namespace detail {
-
-// List of all properties' IDs.
-enum PropKind {
-  // Buffer properties
-  BufferUseHostPtr = 0,
-  BufferContextBound,
-  BufferUseMutex,
-
-  // Image properties
-  ImageUseHostPtr,
-  ImageContextBound,
-  ImageUseMutex,
-
-  // Queue properties
-  QueueEnableProfiling,
-  InOrder,
-
-  // Accessor
-  NoInit,
-
-  PropKindSize
-};
-
-// Base class for all properties. Needed to check that user passed only
-// SYCL's properties to property_list c'tor.
-class PropBase {};
-
-// Second base class, needed for mapping PropKind to class and vice versa.
-template <PropKind PropKindT> class Prop;
+class property_list {
 
-// This class is used in property_list to hold properties.
-template <class T> class PropertyHolder {
-public:
-  PropertyHolder() = default;
+  // Base property wrapper type for type erasure
+  class property_wrapper_base {
+  public:
+    virtual ~property_wrapper_base() = default;
+  };
 
-  PropertyHolder(const PropertyHolder &P) {
-    if (P.isInitialized()) {
-      new (m_Mem) T(P.getProp());
-      m_Initialized = true;
-    }
-  }
+  // Polymorphic wrapper around a concrete property type
+  template <typename T> class property_wrapper : public property_wrapper_base {
+  public:
+    T impl;
 
-  ~PropertyHolder() {
-    if (m_Initialized) {
-      T *MemPtr = reinterpret_cast<T *>(m_Mem);
-      MemPtr->~T();
-    }
-  }
+    property_wrapper(T impl) : impl{std::move(impl)} {}
+  };
 
-  PropertyHolder &operator=(const PropertyHolder &Other) {
-    if (this != &Other) {
-      if (m_Initialized) {
-        T *MemPtr = reinterpret_cast<T *>(m_Mem);
-        MemPtr->~T();
-        m_Initialized = false;
-      }
+  // Map property type indicies against type erased container
+  // Note: Memory fragmentation likely as each property is allocated seperately
+  std::unordered_map<std::type_index, std::shared_ptr<property_wrapper_base>>
```

Comment:
Why `shared_ptr`? Why not `unique_ptr`?

Reply:
`unique_ptr` would have been nice, but since `property_list` can be copied the unordered map will also be copied together with its elements, which `unique_ptr` wouldn't allow.